scan/scan_common.go

```diff
@@ -163,57 +170,72 @@ func (fs FamilySet) RunScans(host, ip, family, scanner string, dur time.Duration
 		return nil, err
 	}
 
-	results := make(chan map[string]FamilyResult)
-	timeout := make(chan bool)
-
-	if dur > 0 {
-		go startTimer(dur, timeout)
-	}
+	results := make(chan *Result)
 
-	familyResults := make(map[string]FamilyResult)
+	var familyWG sync.WaitGroup
+	familyWG.Add(len(fs))
+	go func() {
+		familyWG.Wait()
+		close(results)
+	}()
 
 	go func() {
 		for familyName, family := range fs {
+			var scannerWG sync.WaitGroup
 			if familyRegexp.MatchString(familyName) {
-				scannerResults := make(map[string]ScannerResult)
+				scannerWG.Add(len(family.Scanners))
 				for scannerName, scanner := range family.Scanners {
-					if scannerRegexp.MatchString(scannerName) {
-						grade, output, err := scanner.Scan(addr, hostname)
-						if grade != Skipped {
-							result := ScannerResult{
-								Grade:  grade.String(),
-								Output: output,
-							}
+					go func(scannerName string, scanner *Scanner) {
+						if scannerRegexp.MatchString(scannerName) {
+							grade, output, err := scanner.Scan(addr, hostname)
+							var errStr string
 							if err != nil {
-								result.Error = err.Error()
+								errStr = err.Error()
+							}
+							results <- &Result{
+								Family:  familyName,
+								Scanner: scannerName,
+								ScannerResult: ScannerResult{
+									Grade:  grade.String(),
+									Output: output,
+									Error:  errStr,
+								},
 							}
-
-							scannerResults[scannerName] = result
 						}
-					}
-				}
-
-				if len(scannerResults) > 0 {
-					familyResults[familyName] = scannerResults
+						scannerWG.Done()
+					}(scannerName, scanner)
 				}
 			}
+			go func() {
+				scannerWG.Wait()
+				familyWG.Done()
```

Comment:
This Waits for all the Scanners in a Family to finish before marking it as Done in the WaitGroup.